### ISSUE TITLE
SessionStateUnauthorizedAccessException when Mocking Get-Module

### ISSUE THREAD

**Issue by alx9r** [pester/Pester#1007]
I came across this while trying to repro #1006.

## 1. Repro

```PowerShell
function f { Get-Module foo }

Describe something {
    It 'mock Get-Module' {
        Mock Get-Module -Verifiable { 'mocked' }
        f
        Assert-MockCalled Get-Module
    }
}
```

## 2. Describe Your Environment

Operating System, Pester version and PowerShell version:

```none
Pester version     : 4.3.1 C:\Users\un1\Documents\WindowsPowerShell\Modules\PEster\Pester.psm1
PowerShell version : 5.1.14409.1012
OS version         : Microsoft Windows NT 6.3.9600.0
```

## 3. Expected Behavior

Test should pass. 

## 4.Current Behavior

```none
Describing something
  [-] mock Get-Module 64ms
    SessionStateUnauthorizedAccessException: Cannot overwrite variable PSEdition
 because it is read-only or constant.
    at f, C:\Users\un1\Desktop\test2.ps1: line 1
    at <ScriptBlock>, C:\Users\un1\Desktop\test2.ps1: line 6
```

**Comment by alx9r** [comment]
@dlwyatt Can you take a look at this?  I'm not familiar with Pester's mocking subsystem.

**Comment by nohwnd** [comment]
@alx9r The mock generates implementation of the mock function that has parameter called PSEdition, which causes the error when the function is executed. Mocking internally uses ProxyCommand. Here is a repro that suggests that this is error in the ProxyCommand, not in Pester (yet we should still find a way to  fix it, probably by skipping that parameter). 

```powershell
$MetaData = New-Object System.Management.Automation.CommandMetaData (Get-Command Get-Module) 
$g = [System.Management.Automation.ProxyCommand]::Create($MetaData)
& ([scriptblock]::Create($g))

# Cannot overwrite variable PSEdition because it is read-only or constant.
```

**Comment by alx9r** [comment]
@nohwnd Oh I see now.  Although, I find the details of this behavior to be peculiar:

```PowerShell
& {param($PSEdition)} # fails
& $m { [bool]((Get-Variable PSEdition | % Options) -band 'Constant') } # true
& {param([Alias('PSEdition')]$__PSEdition)} # succeeds
& {param($ExecutionContext)} # fails

Set-Variable IAmAConstant -Value 'constant value' -Option Constant
& {param($IAmAConstant)} # succeeds

$m = New-Module {}
& $m { [bool]((Get-Variable PSEdition | % Options) -band 'Constant') } # true
& $m {param($PSEdition)} # succeeds
```

FWIW, it seems there is discussion around the following in PowerShell/PowerShell#3061:

* which automatic variables should/will become read-only 
* how to discover/detect automatic variables

I'm not convinced that the error correlates well with automatic variables, however.

**Comment by alx9r** [comment]
@nohwnd 

AFAICT, we should expect this error to occur whenever the name of a parameter conflicts with a `ReadOnly` or `Constant` variable in the _current_ scope.  Currently, `Get-Module`'s `$PSEdition` _can_ be mocked inside a module because of a bug that causes `$PSEdition` to be shadowed inside a module despite its `AllScope` option (see PowerShell/PowerShell#6378).  When that bug is resolved, I expect mocking of any parameters that conflict with `AllScope` variables that are `ReadOnly` or `Constant` (like $PSEdition) to fail even inside modules.  In other words, if this bug doesn't get fixed before PowerShell/PowerShell#6378 I _think_ all mocks of `Get-Module` (and probably some others that whose parameters conflict with `ReadOnly` or `Constant` variables) will start failing.

It seems like one possible workaround would be to translate a parameter to use an alias when such a conflict is detected.  For example,

```PowerShell
param(
...
    $PSEdition
...
)
```

would become

```PowerShell
param(
...
    [Alias('PSEdition')]
    $__PSEdition
...
)
```

This and #980 make me wonder whether `[ProxyCommand]::Create()` should be improved or replaced.

**Comment by nohwnd** [comment]
@alx9r I don't think we can do better job than the ProxyCommand without a lot of effort. So I would just add special cases for the errors as they come. The type in #980 can be relaxed to [object] (or string with validation), and the PSEdition can be fixed the way you describe. Ok?

**Comment by alx9r** [comment]
>I don't think we can do better job than the ProxyCommand without a lot of effort.

@nohwnd It looks to me like visiting and emitting a string from each relevant node in the `[Dictionary[string,ParameterMetaData]]` tree would produce the needed `param()` block.  The visitor could implement the special cases from this bug, #980, and whatever else we find.

>So I would just add special cases for the errors as they come. 

I think how well this works will come down to implementation.  If it's string pattern matching and replacement, I would anticipate pernicious edge-cases.

>...PSEdition can be fixed the way you describe.

I haven't put much though into that method.  It might not be the best way.  Using parameter aliases would mean, for example, that the parameter names available in `-ParameterFilter` don't match the real command's signature.

**Comment by nohwnd** [comment]
@alx9r I agree that there might be edge cases, but the language allowed in the param block is quite simple, so doing a string replacement on a known type / parameter name is imho relatively problem free. I would rather start simple, than implement the whole thing ourselves, the proxycommand worked for the past few years without any problems, it's just now that we have 2 bugs in the same year.

**Comment by alx9r** [comment]
> ... but the language allowed in the param block is quite simple, so doing a string replacement on a known type / parameter name is imho relatively problem free.

I think you are probably right for Cmdlets.  I think `param()` for PSCmdlets can contain almost anything in `[ValidationScript()]` scriptblocks, so it might be worth doing the #1007 replacement for only Cmdlets.

> I would rather start simple, than implement the whole thing ourselves, the proxycommand worked for the past few years without any problems, it's just now that we have 2 bugs in the same year.

This sounds reasonable to me.

**Comment by nohwnd** [comment]
This is related to #601 where we want to relax the types. This would remove (rename?) the parameters to avoid conflicts.

**Comment by renehernandez** [comment]
@nohwnd I am going to take a look at this. Per your last comment, the pull request should bring the necessary fixes for both this issue and #601?

**Comment by nohwnd** [comment]
@renehernandez the related issue is way more involved than this one. I’d keep them separate, once you tackle this, you can have a look on the other one as well, but this one is imho a much better starting point for getting familiar with Mock internals :)

**Comment by nohwnd** [comment]
@renehernandez 
> This would remove (rename?) the parameters to avoid conflicts. 

Okay this was a bad idea, removing the parameter would prevent the production code from ever using it, because providing non-existing parameter to a function with `[CmdletBinding()]` makes it fail. So let's rename it and alias it instead, because that works fine. 

```powershell
# gets only the params but no body
# $MetaData = New-Object System.Management.Automation.CommandMetaData (Get-Command Get-Module) 
# $proxy = [System.Management.Automation.ProxyCommand]::Create($MetaData)
# 
# 
# $parameterLines = foreach ($line in ($proxy -split "`n")) {
#     if ($line -like "begin*") {
#         break
#     }
#     $line
# }
# 
# $parameters = $parameterLines -join "`n"
# $parameters

 # define our function with an excerpt from the parameter definiton 
 # we got from proxy, this is pretty much what mock does internally
 Invoke-Expression @'
function Get-Module2 {
    [CmdletBinding(DefaultParameterSetName='Loaded', HelpUri='https://go.microsoft.com/fwlink/?LinkID=141552')]
    param(
        [Parameter(ParameterSetName='Available', Position=0, ValueFromPipeline=$true)]
        [ValidateNotNullOrEmpty()]
        [string[]]
        ${Name},

        [Parameter(ParameterSetName='Available', Mandatory=$true)]
        [Parameter(ParameterSetName='PsSession')]
        [Parameter(ParameterSetName='CimSession')]
        [switch]
        ${ListAvailable},

        # renamed to _PSEdition and alised to the
        # original name
        [Alias('PSEdition')]
        [Parameter(ParameterSetName='Available')]
        [string]
        ${_PSEdition}
     )

    'i am mock'
}
'@

# this is a call that the prodution code would make, removing the PSEdition parameter 
# would fail the call, because the parameter would not exist, so instead of 
# removing the parameter we should rename it and alias it. The rename convention
# is important because that is what needs to be used in ParameterFilter
Get-Module2 -Name "Abc" -ListAvailable -PSEdition "Desktop"
```

**Comment by renehernandez** [comment]
@nohwnd A couple of questions while I work on this:

* Should this apply as well to dynamic parameters? Or is it only applicable to regular parameters?
* What if we apply this alias to every parameter instead of a select few? Wouldn't that cover against future additions of read-only/constant variables?

**Comment by nohwnd** [comment]
> Should this apply as well to dynamic parameters? Or is it only applicable to regular parameters?

I don't recally any issues regarding this in dynamic parameters, right now it should apply only to a list of parameters. I am sure there are others but I don't recally their names. Once the capability is there it will be as simple as adding the problematic parameters to the list.

> What if we apply this alias pattern to every parameter instead of a select few? Wouldn't that cover against future additions of read-only/constant variables?

It would, but at the cost of braking every parameter filter that is out there. So please don't do it. 🙂

**Comment by renehernandez** [comment]
> It would, but at the cost of braking every parameter filter that is out there. So please don't do it. 🙂

I would have expected that I could use a parameter alias in the definition of `ParameterFilter` blocks and it would resolve to right parameter. But it seems that that is not currently the case, as per the follow example:

```
function Test-Params {
    [CmdletBinding()]
    param(
        [Parameter()]
        [Alias('ParamAlias')]
        $Param
    )

    Write-Host $Param
}

Describe "Test-Params" {

    It 'mocks with alias in parameter filter' {
        Mock Test-Params -ParameterFilter { $ParamAlias -eq "Hello" }

        Test-Params -ParamAlias "Hello"

        Assert-MockCalled Test-Params -Exactly 1
    }
}
```
If we use the real parameter name, then the test passes successfully.

So with that in mind, after doing the suggested aliasing of parameters such as *PSEdition*, we wouldn't be able to use them in `ParameterFilters` blocks, because they only work with the real parameter name, not the alias

**Comment by renehernandez** [comment]
@nohwnd I got it working for the provided example for `Get-Module` cmdlet. Currently, I am modifying the `$MetaData` object, by replacing the `ParameterMetadata` that matches with *PSEdition*.  

While doing more testing on the changes, I ran into a failure when the mock doesn't get hit, since it will resort to the original command implementation there and there is no *$_PSEdition* parameter in the original definition, as shown in the test below.

```
function Test-Params {
    [CmdletBinding()]
    param(
        [Parameter()]
        $Param
    )

    $Param
}

Describe "Test-Params" {

    It 'passes when the mock gets hit' {
        Mock Test-Params { "Success" } -ParameterFilter { $_Param -eq "Hello" }

        Test-Params -Param "Hello" | Should -Be "Success"
    }

  It 'fails when the mock does not get hit' {
        Mock Test-Params { "Success" } -ParameterFilter { $_Param -eq "Hello" }

        Test-Params -Param "World" | Should -Be "World" # This fails because the original cmdlet doesn't 
                                                                                         # have $_PSEdition defined
    }
}
```

**Comment by nohwnd** [comment]
@renehernandez hmm that's a good find. 

 I am not sure the call-through would work when the parameters are rewritten in the original command. Is that even possible to do? In never tried.

I would approach this by adding the original and new name to the [mock info](https://github.com/pester/Pester/blob/d7bb364a56877832c8bf783683f568974dfef930/Functions/Mock.ps1#L323-L335), and when it is calling through to the original command I'd modify the PSBoundParameters dictionary to revert the parameters to their original name.
